plugins/single_plugins/preserve-output.cpp

```diff
@@ -0,0 +1,314 @@
+}
+
+/**
+ * preserve-output plugin
+ */
+
+class wayfire_preserve_output : public wf::plugin_interface_t
+{
+    std::vector<std::string> outputs_being_removed;
+
+    wf::signal_connection_t output_pre_remove = [=] (wf::signal_data_t *data)
+    {
+        auto signal_data = (wf::output_pre_remove_signal*)data;
+        LOGD("Received pre-remove event: ", signal_data->output->to_string());
+        outputs_being_removed.push_back(signal_data->output->to_string());
+        LOGD("Outputs currently being removed: ");
+        for (auto el : outputs_being_removed)
+        {
+            LOGD("- ", el);
+        }
+
+        if (signal_data->output != output)
+        {
+            // This event is not for this output
+            return;
+        }
+
+        // This output is being destroyed
+        std::string identifier = make_output_identifier(output);
+
+        // Store this output as the focused one
+        if (wf::get_core().get_active_output()->to_string() == output->to_string())
+        {
+            core_store_focused_output(output);
+        }
+
+        auto views = output->workspace->get_views_in_layer(wf::LAYER_WORKSPACE);
+        for (uint i = 0; i < views.size(); i++)
+        {
+            auto view = views[i];
+            if ((view->role != wf::VIEW_ROLE_TOPLEVEL) || !view->is_mapped())
+            {
+                continue;
+            }
+
+            // Set current output and geometry in the view's last output data
+            if (!view_has_data(view))
+            {
+                view_store_data(view, output, i);
+            }
+        }
+    };
+
+    wf::signal_connection_t output_removed = [=] (wf::signal_data_t *data)
+    {
+        auto signal_data = (wf::output_removed_signal*)data;
+        LOGD("Received output-removed event: ", signal_data->output->to_string());
+        outputs_being_removed.erase(std::remove(
+            outputs_being_removed.begin(),
+            outputs_being_removed.end(),
+            signal_data->output->to_string()), outputs_being_removed.end());
+        LOGD("Outputs currently being removed: ");
+        for (auto el : outputs_being_removed)
+        {
+            LOGD("- ", el);
+        }
+    };
+
+    wf::signal_connection_t restore_views_to_output = [=] (wf::signal_data_t *data)
+    {
+        auto signal_data = (wf::output_added_signal*)data;
+        if (signal_data->output != output)
+        {
+            // This event is not for this output
+            return;
+        }
+
+        std::string identifier = make_output_identifier(output);
+
+        // Focus this output if it was the last one focused
+        if (core_get_focused_output() == identifier)
+        {
+            LOGD("This is last focused output, refocusing: ", output->to_string());
+            wf::get_core().focus_output(output);
+            core_erase_focused_output();
+        }
+
+        for (auto& wo : wf::get_core().output_layout->get_outputs())
```

Comment:
shouldn't you iterate over all views (`wf::get_core().get_all_views()`), build the list, and then sort all the views (instead of doing this per-output?)

Reply:
Yes definitely - I guess it only works because usually there is only one other output with relevant views on them when this code runs.